src/main/scala/com/github/catalystcode/fortis/spark/streaming/reddit/RedditInputDStream.scala

```diff
@@ -0,0 +1,115 @@
+package com.github.catalystcode.fortis.spark.streaming.reddit
+
+import java.util.concurrent.{ScheduledThreadPoolExecutor, TimeUnit}
+
+import com.github.catalystcode.fortis.spark.streaming.reddit.client.RedditClient
+import com.github.catalystcode.fortis.spark.streaming.reddit.dto.RedditObject
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.streaming.StreamingContext
+import org.apache.spark.streaming.dstream.ReceiverInputDStream
+import org.apache.spark.streaming.receiver.Receiver
+
+/**
+  * A RedditInputDStream is a concrete implementation of a {@link ReceiverInputDStream} that provides a {@link Receiver}
+  * instance that is compliant to the Reddit API usage guidelines.
+  *
+  * @see http://github.com/reddit/reddit/wiki/API
+  */
+class RedditInputDStream(val client: RedditClient,
+                         val keywords: Seq[String],
+                         ssc: StreamingContext,
+                         storageLevel: StorageLevel,
+                         val pollingPeriodInSeconds: Int = 2,
+                         val tokenRefreshPeriodInSeconds: Int = (3600/2))
+  extends ReceiverInputDStream[RedditObject](ssc) {
+
+  override def getReceiver(): Receiver[RedditObject] = {
+    logDebug("Creating Reddit receiver")
+    new RedditReceiver(client, keywords, storageLevel, pollingPeriodInSeconds)
+  }
+}
+
+/**
+  * The RedditReceiver is a concrete implementation of a {@link Receiver} that polls the Reddit API according to
+  * Reddit's guidelines, which includes a rate limit of 1 request per 2 seconds.
+  *
+  * @see http://github.com/reddit/reddit/wiki/API
+  */
+private class RedditReceiver(val client: RedditClient,
+                             val keywords: Seq[String],
+                             storageLevel: StorageLevel,
+                             val pollingPeriodInSeconds: Int = 3)
+  extends Receiver[RedditObject](storageLevel) with Logger {
+
+  @volatile private var lastIngestedDate = Long.MinValue
+
+  private var executor: ScheduledThreadPoolExecutor = _
+
+  def onStart(): Unit = {
+    // TODO: Consider implementing using live threads in order to eliminate the need for polling via search.
+    //       https://www.reddit.com/dev/api#GET_live_{thread}
+
+    executor = new ScheduledThreadPoolExecutor(2)
```

Comment:
I see below that we need one thread for the polling and one for the token refresh. It may be worth to make this more explicit, e.g. by having a list of threads and setting the pool capacity to that list's size (or just add a wee comment).